### ISSUE TITLE
How to replace expired or soon to expire CA in Fleet (Elastic Agent & Fleet Server)

### ISSUE THREAD

**Issue by lucabelluccini** [elastic/ingest-docs#167]
We do not provide guidance on how to approach a CA certificates replacement in Fleet Server and/or Elasticsearch with Fleet-managed Elastic Agents.

The CA used by Elastic Agent to trust the Fleet Server cannot be provided in the policy. It is only available as a command line parameter and it points to a local file.
- Do we support multiple CAs in Elastic Agents?
- Can we hot-swap the CA (is it reloaded by Elastic Agent) or it is reloaded only at startup?

Ideally, for updating the CA in Fleet Server without downtime:
- All the Elastic Agents enrolled to the Fleet Server should be updated to trust both the OLD CA and NEW CA. **How?**
- Fleet Server can be restarted, replacing the Fleet Server certificates and CA. **How?**

Ideally, for updating the CA in Elasticsearch without downtime:
- All the Elastic Agents enrolled to the Fleet Server should be updated to trust both the OLD CA and NEW CA. Elasticsearch certs are typically defined in the Fleet UI / Output settings (via reference to a file or embedded in the policy). **How?**
- Fleet Server should be also configured to trust both the OLD CA and NEW CA of Elasticsearch. **How?**
- Elasticsearch should be roll-restarted to update their CA. How to do it it is not in the scope of the guide.

**Comment by leandrojmp** [comment]
Hello, is there any update on this?

We have a requirement to renew the certificates of our Fleet Server that is close to expiration date but I could not find any documentation about this.

**Comment by leandrojmp** [comment]
Hello @lucabelluccini,

Thanks, we already opened a ticket 6 months ago reporting the lack of the documentation about renewing certificates.

At the time we needed to change the CA as well, but this time we fortunately only need to replace the certificate since they are signed by a Know CA.

I opened another ticket today because there isn't yeat any documentation about how to update the certificate before expiring.

Also, it looks to me that this is a product gap as well, having to re-enroll an agent to update the CA is way far from ideal and can be a huge impact for some users.

In my case if I needed to re-enroll thousands of hosts it could take months because of some internal process.

With certificates expiring every year and with the recommendation to expire them every 90 days, the Agent needs to be able to update the certificates and CA more easily.

**Comment by lucabelluccini** [comment]
Hello @leandrojmp - thanks for the additional info.

**Comment by leandrojmp** [comment]
Hello @lucabelluccini,

If it helps, this seems to be the internal issue for an enhancement request that we made last year: https://github.com/elastic/enhancements/issues/20562

**Comment by jlind23** [comment]
@AndersonQ as you recently worked on the mTLS issue, would you be able to assist here?
cc @pierrehilbert

**Comment by nimarezainia** [comment]
Team, can we keep this as the documentation issue. Which is merely to say that we would need a new enrollment with the new certs. This is unfortunately a gap we currently have. Will create another issue for an actual *fix* which will need agent+fleet changes.

**Comment by AndersonQ** [comment]
Hi, let me try to answer the questions:

> Do we support multiple CAs in Elastic Agents?

Yes, the agent can receive a list of CAs

> Can we hot-swap the CA (is it reloaded by Elastic Agent) or it is reloaded only at startup?

It's only loaded when the agent starts.

> All the Elastic Agents enrolled to the Fleet Server should be updated to trust both the OLD 
CA and NEW CA. How?

It's technically possible, but right now defining and loading certificates from the policy is being done for mTLS, which does not work on cloud. tl;dr: some development is needed, but with  https://github.com/elastic/elastic-agent/issues/2247 and the mTLS as a whole, most of the work should be done. Then it's be just to add both CAs in the policy and the agent would use them both to verify the certificate presented by fleet-server.

> Fleet Server can be restarted, replacing the Fleet Server certificates and CA. How?

If the certificates are defined as a path, it should be possible to replace the contents of the file and fleet server should pick up the new ones. As far as I remember it isn't possible to load multiple certificates, only multiple CAs. 

I'm not sure, but I believe the agent does not support receiving a directory for CAs or certificates. If it was supported, the contents of the directory could be changed, adding or removing certificates and CAs, and on restart the agent would pick up the updated certificates/CAs

**Comment by kilfoyle** [comment]
Thanks @AndersonQ! This helps a lot. Once the mTLS work lands it seems the process will be a lot simpler.

@nimarezainia We have this Secure connections section in the docs, but since that section is only about on-prem setups I would lean toward putting this content under "Manage Elastic Agents in Fleet" -> "[Elastic Agents](https://www.elastic.co/guide/en/fleet/current/manage-agents.html)".

**Comment by nimarezainia** [comment]
> I'm not sure, but I believe the agent does not support receiving a directory for CAs or certificates. If it was supported, the contents of the directory could be changed, adding or removing certificates and CAs, and on restart the agent would pick up the updated certificates/CAs

I'm fairly certain that this is supported but am struggling to find reference to it in docs/issues :-( We had the same concern for Logstash and trying to find a way to send each agent individual certificates. I agree that it may not be valid for CA.

**Comment by nimarezainia** [comment]
@AndersonQ not what I was looking for but in our docs [HERE](https://www.elastic.co/guide/en/fleet/current/secure-connections.html#generate-fleet-server-certs) you can the see examples where CA and Cert can be loaded from a directory. I guess the question is whether agent/fleet-server read from this path on every new handshake (or do they rely on what's in memory):

sudo elastic-agent install --url=https://192.0.2.1:8220 \
  --enrollment-token=<string> \
  --certificate-authorities=/path/to/ca.crt

sudo ./elastic-agent install \
   --url=https://192.0.2.1:8220 \
   --fleet-server-es=https://192.0.2.0:9200 \
   --fleet-server-service-token=AAEBAWVsYXm0aWMvZmxlZXQtc2XydmVyL3Rva2VuLTE2MjM4OTAztDU1OTQ6dllfVW1mYnFTVjJwTC2ZQ0EtVnVZQQ \
   --fleet-server-es-ca=/path/to/elasticsearch-ca.crt \
   --certificate-authorities=/path/to/ca.crt \
   --fleet-server-cert=/path/to/fleet-server.crt \
   --fleet-server-cert-key=/path/to/fleet-server.key \
   --fleet-server-port=8220

**Comment by nimarezainia** [comment]
When you install an on-prem fleet server:

<img width="706" alt="image" src="https://github.com/elastic/ingest-docs/assets/13442829/e92e85ab-7a7e-4966-8e20-a2ae64f034d8">


as @cmacknz mentioned in the other issue we don't expose these tls configs in the UI however the path to CA and Cert can be configured for the fleet server. Queston remains when these get read.

**Comment by AndersonQ** [comment]
@nimarezainia when I said

> I believe the agent does not support receiving a directory for CAs or certificates.

I meant being able to pass a directory containing several CAs/certificates and the agent loading them all, like the [`--capath`](https://curl.se/docs/manpage.html#--capath) of cURL.

Sorry if I wasn't clear on that.

**Comment by mathy-ufm** [comment]
Has anybody been able to change the CA and certificate on the fleet server, aswell as on the clients without re-enrolling them ? Also.. How do we change them in the policies ? Even though Ive changed the certificates they are pointing to,
![fleet_policy](https://github.com/elastic/ingest-docs/assets/146924211/3728c456-ce29-4e91-86bf-5dcea5e9cea2)
 the old one still shows when I look at the policy....

Brgds.

**Comment by leandrojmp** [comment]
@mathy-ufm if you need to change the CA that you are using in your fleet server, then there is no way to change it without re-enrolling all your agents.

The CA for the fleet server is set when you run the install command, after that you cannot change, so if your CA expired or you need to change it, you will need to reinstall the fleet server with the new CA and doing that you will need to re-enroll your agents.

**Comment by kilfoyle** [comment]
Just coming back to this one: We've documented certificate rotation in [Rotate SSL/TLS CA certificates](https://www.elastic.co/docs/reference/fleet/certificates-rotation), which covers:

* Rotating a Fleet Server CA
* Rotating an Elasticsearch CA for connections from Fleet Server
* Rotating an Elasticsearch CA for connections from Elastic Agent

I don't think that those sections cover all of the questions in the issue description, such as support for multiple CAs, whether or not CA hot-swapping is supported, etc.

**Comment by lucabelluccini** [comment]
Hello, 
I think the docs https://www.elastic.co/docs/reference/fleet/certificates-rotation are a great addition.

For the part Elastic Agent > Fleet Server, I think that in recent versions we should be able to change the SSL settings also from Fleet (I think 9.1 introduced the SSL settings also for the Fleet Server connection, which would override the one at bootstrap time).

> The CA for the fleet server is set when you run the install command, after that you cannot change, so if your CA expired or you need to change it, you will need to reinstall the fleet server with the new CA and doing that you will need to re-enroll your agents.

If we're talking about the CA used to connect from the Fleet Integration to Elasticsearch, I think starting 8.15 the ES settings in the policy can override the ones set at enrollment time.

---

I think users are mostly interested into knowing how to rotate the CA and keep everything running, without re-enrolling and whenever possible, without using commands which would require 3rd party configuration management tools (e.g. puppet, chef, etc...). Ideally, we should aim to allowing this kind of operations purely via Fleet UI. But I do not know if we already did a study of all the connection requirements and our ability to change the settings without interruptions/downtime.
